### ISSUE TITLE
Exiting Output-ConsoleGridView GUI is too hard

### ISSUE THREAD

**Issue by tig** [PowerShell/GraphicalTools#61]
To exit the user has to press F9 then accept the confirmation.

Is there a reason this was made so multi-step? 

Why not just have a keyboard shortcut (like ESC)?

**Comment by TylerLeonhardt** [comment]
It's because I copied from an example 😅I'm not apposed to a single click button or single keyboard shortcut at all.

**Comment by potatoqualitee** [comment]
Escape would be great to escape and then perhaps a double enter to make a selection? Or let's see how suse does it.  Accept down at the lower right similar to GUI with no confirmation.

**Comment by gwojan** [comment]
I think the default should be just like how `Out-GridView` works...

**Comment by tig** [comment]
Yeah. `Out-GridView` uses ESC to exit and ENTER to send selected items `out`. 

Imma gonna look into doing it... Hold my beer.

**Comment by corbob** [comment]
The way I would expect it to work would be: `ESC` returns nothing. If you're selecting multiple items, you select them with `Space`, and finally `Enter` would return whatever is selected (or under the cursor if nothing is selected).

**Comment by tig** [comment]
Here's what I have working. I like it; seems very natural:

````
                // Regardless of whether the item was already marked, if the user presses
                // ENTER, mark that item. This feels the most intuitive. The alternatives are:
                // (a) Assume the user means to choose the item that's selected and mark it
                // (b) Act like ESC except Pass Through an marked items.
                // (c) Toggle the selected item
                // Option (c) violates the principle of least astonishment: If the user has an item marked, and selected
                // and hits ENTER s/he would be surprised to see it not returned. Esp in the single choice (most common) case.
                // Option (b) means in the most common case (a single item is wanted) ENTER does nothing. Again surprising.
                // It seems a) is the most intuitive so that's what's implemented.
```

**Comment by potatoqualitee** [comment]
Boooooom so fast! Rad

**Comment by TylerLeonhardt** [comment]
ESC is tough because it's typically used to close out of menus and whatnot. My ideal option would be:

* With PassThru (which will probably be the default behavior soon #71), two buttons at the bottom:
  * `Accept (<some keybinding>)` & `Close (<some keybinding>)`.
* Without PassThru, two buttons at the bottom:
  * `Close (<some keybinding>)`

 These would be buttons you could click or use a keybinding to trigger that way we have both mouse and keyboard scenarios handled.

I would prefer that keybindings be at least 2 keys or a function key that way you can be absolutely sure you mean what you type.


Keep in mind, I'm doing some refactor work in #74 so we should hold off on implementations until that's done.

**Comment by TylerLeonhardt** [comment]
A good first step to shave off some keyboard strokes is to keep the menu we already have and just delete the `Quit` method code:

https://github.com/PowerShell/GraphicalTools/blob/0406f50c1ca2bdff80917cbf979cbb4f1e7f56de/src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs#L61

**Comment by gwojan** [comment]
Honestly, @TylerLeonhardt what's your objection to using the same keybindings as `Out-GridView`? It just seems so natural to me -- especially in passthru mode.

**Comment by tig** [comment]
> Honestly, @TylerLeonhardt what's your objection to using the same keybindings as `Out-GridView`? It just seems so natural to me -- especially in passthru mode.

On Windows the keybinding to exit `out-gridview` is `alt-F4`. Not sure what it is on mac.

`alt-F4` in a Windows Terminal session will close it, not the contained `out-consolegridview`.

On Mac and Windows, `enter` is the standard keyboard idiom for exiting (accepting) a modal dialog. A console Cmdlet is a modal dialog. 

On Mac and Windows, `esc` is the standard keyboard idiom for cancelling a modal dialog. A console Cmdlet is a modal dialog. 

Are there very many other Powershell cmdlets that are doing `gui.c` stuff? What norms are they following? 

I choose ESC/ENTER because it matches what happens with other modal dialogs and is simple for the thing that `out-consolegridview` does. If `-passthru` were default it'd be even more natural:

* As a user I want to select one item from a list in a console grid view with as few keystrokes as possible (without using the mouse). --> <ARROWS> ENTER
* As a user I want to select multiple items from a list in a console grid view with as few keystrokes as possible (without using the mouse) --> <ARROWS> SPACE <ARROWS> SPACE ENTER
* As a user I want to abort a selection using console grid view with as few keystrokes as possible --> ESC

The menu is required only for mouse usage. There it shouldn't be a menu, but just 'buttons' like a modal dialog as @TylerLeonhardt suggesed with:

> With PassThru (which will probably be the default behavior soon #71), two buttons at the bottom:

> Accept (<some> ENTER>) & Close (<some ESC>).

**Comment by TylerLeonhardt** [comment]
Yeah ok that sounds good. The refactor work is done for now so feel free to send a new PR!

I'd say the PR should either:
* Just add the keybindings and leave the menu as it is
or
* Add the keybindings and replace the menu with buttons

**Comment by TylerLeonhardt** [comment]
The only problem I have with that proposal is that the `-OutputMode` parameter is suppose to control that single ENTER behavior:

https://github.com/PowerShell/GraphicalTools/issues/79

**Comment by tig** [comment]
I don't think #79 will be a problem. 

I now have things working with your new code, however it depends on the new `StatusBar` that is now in `gui.cs` as of version `0.80.0`. 

https://github.com/migueldeicaza/gui.cs/blob/49cd29853f5de55547c9011996ed982040607b5c/Terminal.Gui/Views/StatusBar.cs

But version `0.80 .0` has not dropped to nuget yet.

`StatusBar` makes this all so much better:



```powershell
PS> gci | Out-ConsoleGridView -PassThru
```

![image](https://user-images.githubusercontent.com/585482/79092343-daf8b280-7d0d-11ea-9eb1-45db850fa278.png)

```powershell
gci | Out-ConsoleGridView
```

![image](https://user-images.githubusercontent.com/585482/79092388-f82d8100-7d0d-11ea-938b-f20a3dd6e065.png)

Implementing `-OutputMode Single` will simply require the item selection code to only allow one item to be connected, I think. `-OutputMode Multiple` is identical to `-PassThru`.

Side note: In the above, I've disabled your filter code because it's exposing a bug in `gui.cs` `0.80.0` (I think), where `Pos.Right(...)` calls are causing an invalid cast exception deep in `gui.cs`. You'll see this as soon as you try to use `0.80.0`.

**Comment by tig** [comment]
FWIW, I found where the bug in `gui.cs` is that was causing the filter code to crash:

https://github.com/migueldeicaza/gui.cs/issues/356

Looks nice!

![image](https://user-images.githubusercontent.com/585482/79094114-ae479980-7d13-11ea-8baa-e0a720806b23.png)

Workaround is this:

```csharp
            // Workaround for https://github.com/migueldeicaza/gui.cs/issues/356
            filterFieldWidth += 2;
            var filterApplyButton = new Button(APPLY_LABEL)
            {
                // Pos.Right(filterField) returns 0
                X = Pos.Right(filterLabel) + filterFieldWidth,
                Y = Pos.Top(filterLabel),
                Clicked = () =>
                {
                    filterChanged.Invoke(null, filterField.Text);
                }
            };
```

**Comment by TylerLeonhardt** [comment]
looks a lot nicer in Windows Terminal too!

**Comment by TylerLeonhardt** [comment]
Are Accept and Close clickable?

**Comment by TylerLeonhardt** [comment]
My only concern is that Space to _select_ is not very discoverable.

**Comment by tig** [comment]
> Are Accept and Close clickable?

Not yet (`StatusBar` doesn't support that). I've added it as an issue to `gui.cs`: https://github.com/migueldeicaza/gui.cs/issues/360

**Comment by TylerLeonhardt** [comment]
Seems like we should wait until `StatusBar` is a bit more stable (and released 😅) before moving to that.

We can keep the top menu for now - but support the keybindings you've suggested.
Or
we can wait on gui.cs to get StatusBar out but from experience, the project does take time to do releases.


This is awesome work @tig. I'm a big fan!

**Comment by TylerLeonhardt** [comment]
This is fixed now pending a new release
